File: apps/api/src/tradingos_api/schemas/recommendations.py
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel

from tradingos_api.models.enums import (
    AgentRole,
    AgentRunStatus,
    CommitteeSessionStatus,
    RecommendationAction,
    RecommendationConfidence,
    RecommendationLevelKind,
    RecommendationStatus,
)
from tradingos_api.schemas.instruments import InstrumentResponse


class RecommendationLevelResponse(BaseModel):
    kind: RecommendationLevelKind
    price: Decimal
    basis: str | None

    model_config = {"from_attributes": True}


class RecommendationVersionResponse(BaseModel):
    id: uuid.UUID
    version_number: int
    action: RecommendationAction
    confidence: RecommendationConfidence
    score: Decimal | None
    rationale: str
    generated_at: datetime
    levels: list[RecommendationLevelResponse]
    committee_session_id: uuid.UUID | None = None
    """Revision Prompt 15 — the one-click path from a recommendation to
    its committee's per-role opinions (`GET .../committee-sessions/{id}`).
    Read-only exposure of an already-existing column; adds no new
    computation and changes no trading decision."""


class RecommendationSummaryResponse(BaseModel):
    id: uuid.UUID
    instrument: InstrumentResponse
    status: RecommendationStatus
    opened_at: datetime
    latest_version: RecommendationVersionResponse | None


class AgentOpinionResponse(BaseModel):
    stance: str | None
    structured_output: dict[str, Any]

    model_config = {"from_attributes": True}


class AgentRunResponse(BaseModel):
    id: uuid.UUID
    role: AgentRole
    status: AgentRunStatus
    started_at: datetime | None
    completed_at: datetime | None
    opinion: AgentOpinionResponse | None


class CommitteeSessionDetailResponse(BaseModel):
    id: uuid.UUID
    instrument: InstrumentResponse
    triggered_by: str
    status: CommitteeSessionStatus
    started_at: datetime | None
    completed_at: datetime | None
    agent_runs: list[AgentRunResponse]

File: apps/web/lib/api/recommendations.ts
import { apiGet } from "./client";

export type RecommendationLevel = {
  kind: "ENTRY" | "STOP" | "TARGET";
  price: string;
  basis: string | null;
};

export type RecommendationVersionDetail = {
  id: string;
  version_number: number;
  action: string;
  confidence: "LOW" | "MEDIUM" | "HIGH";
  score: string | null;
  rationale: string;
  generated_at: string;
  levels: RecommendationLevel[];
  committee_session_id: string | null;
};

export type RecommendationDetail = {
  id: string;
  instrument: { id: string; ticker: string; name: string | null };
  status: string;
  opened_at: string;
  latest_version: RecommendationVersionDetail | null;
};

export function getRecommendation(id: string): Promise<RecommendationDetail> {
  return apiGet<RecommendationDetail>(`/api/v1/recommendations/${id}`);
}

/**
 * `MorningPlanItem.recommendation_version_id` (docs/MORNING_PLAN_SPEC.md's
 * version-scoped job-lineage manifest) is a `RecommendationVersion` id,
 * not a `Recommendation` id — this resolves one directly rather than
 * requiring a caller to already know the parent recommendation.
 */
export function getRecommendationVersion(versionId: string): Promise<RecommendationVersionDetail> {
  return apiGet<RecommendationVersionDetail>(`/api/v1/recommendations/versions/${versionId}`);
}

export type AgentOpinion = {
  stance: string | null;
  structured_output: Record<string, unknown>;
};

export type AgentRun = {
  id: string;
  role: string;
  status: string;
  started_at: string | null;
  completed_at: string | null;
  opinion: AgentOpinion | null;
};

export type CommitteeSessionDetail = {
  id: string;
  instrument: { id: string; ticker: string; name: string | null };
  triggered_by: string;
  status: string;
  started_at: string | null;
  completed_at: string | null;
  agent_runs: AgentRun[];
};

/**
 * "Committee disagreement" (Revision Prompt 15's one-click-access
 * requirement) is derived client-side purely for display — counting
 * distinct non-null stances among `agent_runs` — never used to compute
 * or alter a trading decision, which the backend already finalized when
 * it wrote the `RecommendationVersion`.
 */
export function getCommitteeSession(sessionId: string): Promise<CommitteeSessionDetail> {
  return apiGet<CommitteeSessionDetail>(`/api/v1/recommendations/committee-sessions/${sessionId}`);
}
